lib/analysis/author/community.py

```diff
@@ -0,0 +1,265 @@
+            output_file.write(email_addr+","+",".join([str(x) for x in data_list])+"\n")
+        output_file.close()
+    print("Authors data written to file.")
+
+
+def write_pajek(author_graph, filename="author_graph.net"):
+    # Write Pajek file compatible with the Infomap Community Detection module
+    nx.write_pajek(author_graph, filename)
+    lines_in_file= list()
+    with open(filename, 'r') as pajek_file:
+        for line in pajek_file:
+            lines_in_file.append(line)
+    num_vertices = int(lines_in_file[0].split()[1])
+    for i in range(1, num_vertices+1):
+        line = lines_in_file[i].split()
+        line[1] = "\"" + line[1] + "\""
+        del line[2:]
+        line.append("\n")
+        lines_in_file[i] = " ".join(line)
+    with open(filename, 'w') as pajek_file:
+        for line in lines_in_file:
+            pajek_file.write(line)
+    print("Written to:", filename)
+
+
+def write_pajek_for_submodules(json_data, tree_filename="infomap/output/"+"author_graph.tree"):
+    """
+
+    :param tree_filename:
+    :param json_data:
+    :return:
+    """
+    current_module = 1
+    authors_in_module = set()
+    with open(tree_filename, 'r') as tree_file:
+        for line in tree_file:
+
+            if line[0] == '#':
+                continue
+
+            if int(line[:line.index(":")]) > current_module:
+                author_graph = nx.DiGraph()
+                for msg_id, message in json_data.items():
+                    if message['Cc'] is None:
+                        addr_list = message['To']
+                    else:
+                        addr_list = message['To'] | message['Cc']
+                    # Adding only the required edges to the authors graph:
+                    for to_address in addr_list & authors_in_module:
+                        if author_graph.has_edge(message['From'], to_address):
+                            author_graph[message['From']][to_address]['weight'] += 1
+                        else:
+                            author_graph.add_edge(message['From'], to_address, weight=1)
+                output_filename = "submodule_"+str(current_module)+".net"
+                write_pajek(author_graph, filename=output_filename)
+                # Run the infomaps algorithm
+                output_folder = 'output_submodule' + str(current_module) + "/"
+                subprocess.run(args=['mkdir', output_folder])
+                subprocess.run(args=['./infomap/Infomap', output_filename + ' ' + output_folder
+                               +' --tree --bftree --btree -d -c --node-ranks --flow-network --map'])
+
+                current_module += 1
+                authors_in_module = {line[line.index("\"")+1:line.rindex("\"")]}
+            else:
+                authors_in_module.add(line[line.index("\"")+1:line.rindex("\"")])
+
+
+def vertex_clustering(json_filename, nodelist_filename, edgelist_filename, foldername, time_limit=None, ignore_lat=False):
+    """
```

Comment:
sphinx docs please.

Reply:
Please check b8193f4c7935c85e262dfae192c1b2d30abbec7d for documentation on community module